File: VTB.Sourcebook/Shared/QueryModel/Query.cs
namespace VTB.Sourcebook.QueryModel;

public static class Query
{
    private static readonly QueryNode emptyNode = new EmptyNode();
    
    public static QueryNode Literal(object value)
    {
        if (value is LiteralNode lit)
            return lit;
        else if (value is string)
            return new LiteralNode(LiteralType.String, value);
        else if (value is int)
            return new LiteralNode(LiteralType.Number, value);
        else if (value is DateTime)
            return new LiteralNode(LiteralType.Date, value);
        else if (value is bool b)
            return new LiteralNode(LiteralType.Number, b ? 1 : 0);
        else
            throw new SourcebookException($"Invalid literal type: \"{value.GetType().Name}\".");
    }

    public static QueryNode Name(string name) => new NameNode(name);

    public static QueryNode Not(QueryNode node) => new NotNode(node);

    public static QueryNode Equals(string left, object right) =>
        Binary(BinaryOperator.Equals, Name(left), Literal(right));

    public static QueryNode NotEquals(string left, object right) =>
        Binary(BinaryOperator.NotEquals, Name(left), Literal(right));

    public static QueryNode Lesser(string left, object right) =>
        Binary(BinaryOperator.Lesser, Name(left), Literal(right));

    public static QueryNode Greater(string left, object right) =>
        Binary(BinaryOperator.Greater, Name(left), Literal(right));

    public static QueryNode LesserOrEquals(string left, object right) =>
        Binary(BinaryOperator.LesserOrEquals, Name(left), Literal(right));

    public static QueryNode GreaterOrEquals(string left, object right) =>
        Binary(BinaryOperator.GreaterOrEquals, Name(left), Literal(right));

    public static QueryNode Like(string left, string right) =>
        Binary(BinaryOperator.Like, Name(left), Literal(right));

    public static QueryNode Like(string left, LiteralNode right) =>
        Binary(BinaryOperator.Like, Name(left), right);

    public static QueryNode Binary(BinaryOperator op, QueryNode left, QueryNode right)
    {
        if (left is null || right is null)
            throw new SourcebookException("Invalid node value.");

        return new BinaryNode(op, left, right);
    }

    public static QueryNode Between(string field, DateTime lowerBound, DateTime upperBound) =>
        Between(Name(field), Literal(lowerBound), Literal(upperBound));

    public static QueryNode Between(QueryNode node, QueryNode lowerBound, QueryNode upperBound)
    {
        if (node is null || upperBound is null || lowerBound is null)
            throw new SourcebookException("Invalid node value.");

        return new BetweenNode(node, lowerBound, upperBound);
    }

    public static QueryNode And(QueryNode left, QueryNode right)
    {
        if (left is null || right is null)
            throw new SourcebookException("Invalid node value.");

        return new AndNode(left, right);
    }

    public static QueryNode And(params QueryNode[] nodes)
    {
        if (nodes?.Length < 2)
            throw new SourcebookException("Invalid node value.");

        QueryNode? prev = null;

        foreach (var n in nodes!)
        {
            if (prev is null)
                prev = n;
            else if (prev is AndNode a)
                prev = new AndNode(a, n);
            else
                prev = new AndNode(prev, n);
        }

        return prev!;
    }

    public static QueryNode Or(QueryNode left, QueryNode right)
    {
        if (left is null || right is null)
            throw new SourcebookException("Invalid node value.");

        return new OrNode(left, right);
    }

    public static QueryNode Or(params QueryNode[] nodes)
    {
        if (nodes?.Length < 2)
            throw new SourcebookException("Invalid node value.");

        QueryNode? prev = null;

        foreach (var n in nodes!)
        {
            if (prev is null)
                prev = n;
            else if (prev is OrNode a)
                prev = new OrNode(a, n);
            else
                prev = new OrNode(prev, n);
        }

        return prev!;
    }

    public static QueryNode Empty() => emptyNode;
}

File: web-ui/src/helpers/query.ts
/**
 * Динамические запросы
*/
import { QueryRequest } from '~/shared/models/query-request'; 
/**
 * Функция сложение/вычетание база
 * @param type Тип
 * @param arr Массив
 * @returns 
 */

export function andOr(type: string, arr: Array<any>):QueryRequest{
    if (arr.length == 1 && Array.isArray(arr[0]))
        arr = arr[0];

    var prev = null;
        
    for (var i = 0; i < arr.length; i++) {
        if (prev == null)
            prev = arr[i];
        else if (prev["$type"] == type)
            prev = {
                "$type": type,
                "left": prev,
                "right": arr[i]
            };
        else
            prev = {
                "$type": type,
                "left": prev,
                "right": arr[i]
            };
    }

    return prev;
};

export function and(arr: Array<any>):QueryRequest{
    return andOr("AndNode", arr)
};

export function qName(name:string):QueryRequest {
    return {
        "$type": "NameNode",
        "name": name
    }
};

export function literal(value:string, type:string):QueryRequest {
    return {
        "$type": "LiteralNode",                        
        "type": type,
        "value": value,
    }
}

export function binaryNode(name:string, operator:string, value:any, type:string | null) :QueryRequest{
    return {
        "$type": "BinaryNode",
        "operator": operator,
        "left": qName(name),
        "right": literal(value, type ?? "String")
    }
};

export function equals(name: string, value: string, type: string) :QueryRequest{
    return binaryNode(name, "Equals", value, type);
}

export function like(name: string, value: string, type: string):QueryRequest {
    return binaryNode(name, "Like", value, type);
}
